### PR TITLE
Feature/56

### PR DESCRIPTION
**Pull request by shivenduverma-sap**

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -8,6 +8,12 @@ replace github.com/opencontainers/go-digest => github.com/opencontainers/go-dige
 
 replace github.com/google/go-containerregistry => github.com/google/go-containerregistry v0.14.1-0.20230409045903-ed5c185df419
 
+replace k8s.io/apimachinery => k8s.io/apimachinery v0.27.3
+
+replace k8s.io/api => k8s.io/api v0.27.3
+
+replace k8s.io/apiextensions-apiserver => k8s.io/apiextensions-apiserver v0.26.3
+
```
Comment (Skarlso) [inline go.mod:16]: ?

File: hack/create_developer_certificate_secrets.sh [hunk 2]
```diff
@@ -0,0 +1,50 @@
+#!/usr/bin/env bash
```
Comment (Skarlso) [inline hack/create_developer_certificate_secrets.sh:1]: The ocm controller is responsible for creating these. Then they have to be applied and used by other controllers. Why was this added here?

File: e2e/suite_test.go [hunk 3]
```diff
@@ -40,18 +42,18 @@ func TestMain(m *testing.M) {
...
 	)
 
 	testEnv.Finish(
-		shared.RemoveGitServer(namespace),
+		//shared.RemoveGitServer(namespace),
 		shared.ShutdownPortForward(stopChannelRegistry),
 		shared.ShutdownPortForward(stopChannelGitea),
-		envfuncs.DeleteNamespace(namespace),
-		envfuncs.DestroyKindCluster(kindClusterName),
+		//envfuncs.DeleteNamespace(namespace),
```
Comment (Skarlso) [inline e2e/suite_test.go:54]: These should be re-enabled.

File: .github/workflows/e2e.yaml [hunk 4]
```diff
@@ -57,9 +89,65 @@ jobs:
           go work init
           go work use mpas
           go work use ocm-controller
+          go work use mpas-product-controller
+          go work use mpas-product-controller
```
Comment (Skarlso) [inline .github/workflows/e2e.yaml:93]: I assume one of these is meant to be project?

File: Makefile [hunk 5]
```diff
@@ -37,3 +60,18 @@ release-bootstrap-component:
...
 test:
 	go test -v ./internal/... $(GO_TEST_ARGS) -coverprofile cover.out
+
+# https registry
+MKCERT_VERSION ?= v1.4.4
+MKCERT ?= $(LOCALBIN)/mkcert
+UNAME ?= $(shell uname|tr '[:upper:]' '[:lower:]')
+
+.PHONY: generate-developer-certs
+generate-developer-certs: mkcert
```
Comment (Skarlso) [inline Makefile:70]: These should use ocm-controller's scripts and not replicate it. It will cause a drift if there is any chance in constructing certificates.

File: .github/workflows/e2e.yaml [hunk 6]
```diff
@@ -28,7 +46,7 @@ jobs:
         uses: actions/checkout@v3
         with:
           repository: ${{ github.repository }}
-          ref: main
+          ref: feature/56
```
Comment (Skarlso) [inline .github/workflows/e2e.yaml:49]: TODO: change this.

File: go.mod (hunk 1 above)
Comment (shivenduverma-sap) [reply go.mod:16]: The latest versions v.0.27.4 requires changes to use them, replaced them with the older version to not fix in this ticket

File: hack/create_developer_certificate_secrets.sh (hunk 2 above)
Comment (shivenduverma-sap) [reply hack/create_developer_certificate_secrets.sh:1]: When running make e2e from the ocm-controller, these are created in the ocm-controller and it creates the secret in the ocm-system namespace

when running make e2e from mpas in the action, these certs are not available and not created in the mpas-system namespace

File: go.mod (hunk 1 above)
Comment (Skarlso) [reply go.mod:16]: Was there any kind of error anywhere?

Comment (shivenduverma-sap) [reply go.mod:16]: Yes, the new versions have a cache implementation which requires a few changes

Comment (Skarlso) [reply go.mod:16]: Can you point me to that code change?

File: hack/create_developer_certificate_secrets.sh (hunk 2 above)
Comment (Skarlso) [reply hack/create_developer_certificate_secrets.sh:1]: So... copy them over?

File: go.mod (hunk 1 above)
Comment (shivenduverma-sap) [reply go.mod:16]: Eg. Where ever you have: &source.Kind{Type: 
Replace with NewKindWithCache() in controller-runtime

And there were more, feel free to test with latest versions and check the failure messages

File: hack/create_developer_certificate_secrets.sh (hunk 2 above)
Comment (shivenduverma-sap) [reply hack/create_developer_certificate_secrets.sh:1]: You have to assume that you 'ocm-controller/hack/certs' is empty because the git does not have certificates committed for the action to use. Hence the ocm-controller/Tiltfile will be unable to create the secret

Also, if someone is running the mpas e2e tests via make e2e, they should not require to goto the ocm-controller and run make generate-developer-certs which should generate certs and copy over to mpas-system namespace

Comment (shivenduverma-sap) [reply hack/create_developer_certificate_secrets.sh:1]: Correction:

> When running make e2e from the ocm-controller, these are created in the ocm-controller and it creates the secret in the ocm-system namespace

When running make e2e from the ocm-controller, certificates are created in the ocm-controller project

> when running make e2e from mpas in the action, these certs are not available and not created in the mpas-system namespace
when running make e2e from mpas in the action, these certs are not available and not created in the mpas project

File: Makefile (hunk 5 above)
Comment (shivenduverma-sap) [reply Makefile:70]: The roduct-controller Tiltfile replaces the ocm-system registry secret if it is created as well with the secret being the same in mpas-system and ocm-system. There should be no drift

Why mpas creates certs is explained here: https://github.com/open-component-model/MPAS/pull/68#discussion_r1289671064

File: go.mod (hunk 1 above)
Comment (Skarlso) [reply go.mod:16]: Sorry, where is that cache change? I'm not finding it. :D Can you point me to where the code change is in our controllers that need these replace statements? I assume some compiler error was there or something like that?

Comment (shivenduverma-sap) [reply go.mod:16]: There is no code change in our controllers
In order to support the latest versions, we need to adjust a few things

Yes, there were compiler errors

The replace is downgrading from the latest version

File: hack/create_developer_certificate_secrets.sh (hunk 2 above)
Comment (Skarlso) [reply hack/create_developer_certificate_secrets.sh:1]: MPAS e2e is running ocm-controller. Ocm controller has the right to generate certificates. However is running mpas tests should require ocm controller to generate the certs since the ocm controller is the one launching a service that requires the certs.

When running e2e from mpas, the first step is cloning all the controllers, and the next step should be generate the developer certificates. Since the controllers in this namespace also require these certificates, apply them with a patched namespace through tilt into the MPAS namespace.

File: go.mod (hunk 1 above)
Comment (Skarlso) [reply go.mod:16]: Ah, so you run a `go get -u` or something?

File: hack/create_developer_certificate_secrets.sh (hunk 2 above)
Comment (shivenduverma-sap) [reply hack/create_developer_certificate_secrets.sh:1]: And why should we follow that workflow when the 2 workflows can be decoupled? I don't see a reason for this complicated to and fro
the bottom line is, you need authentication secrets and you generate them where the test is running. Unnecessary complexity is not really required for tests. It also mimics what a user has to do, we will ask them to refer to the e2e-tests for understanding the flow. They should not need to understand how all 9 projects are working and which tiltfile is doing what

We will have even more services later on (vault, prometheus, etc.) That doesn't mean we do the ping pong always in different projects and maintain this silo

If you still disagree, I suggest you can create a follow up ticket for improvements

File: go.mod (hunk 1 above)
Comment (shivenduverma-sap) [reply go.mod:16]: All dependencies are updated to the latest version if I run a go get -u or go mod tidy

Comment (Skarlso) [reply go.mod:16]: `go mod tidy` does _not_ update any dependencies. It fixes sha entries, removes or adds missing information. It does not update packages.

File: hack/create_developer_certificate_secrets.sh (hunk 2 above)
Comment (Skarlso) [reply hack/create_developer_certificate_secrets.sh:1]: >the bottom line is, you need authentication secrets and you generate them where the test is running.

That is incorrect. It's a certificate, not an authentication secret.

>It also mimics what a user has to do, we will ask them to refer to the e2e-tests for understanding the flow.

Normally, a user would use cert-manager to deploy proper secrets and not a development flow.

>They should not need to understand how all 9 projects are working and which tiltfile is doing what

They will have to if they want to work on MPAS.

**Comment by Skarlso** [comment]
Updated the PR to use the ocm-controller certificate creation. Also, removed the receiver and adding the project description to the repository which is not required.

I also fixed using the local Tiltfile that encodes applying the right generated certificate.

The test is now failing with some missing files which I assume you'll add @shivenduverma-sap.

```
=== Failed
=== FAIL: e2e TestMpasE2e (0.00s)
    main_test.go:41: running mpas happy path tests
    main_test.go:53: failed to read config file: %w open testdata/podinfo/backend/config.yaml: no such file or directory
```

**Comment by Skarlso** [comment]
After adding back the deleted files, the tests seem to work expect for a couple other small problems.

```
FAIL: TestMpasE2e/Create_Project/1.12_check_files_are_created_in_project_repo (0.05s)

```

This failed because some `deploy.yaml` file was also missing. Probably some deleted things that need to be updated.

**Comment by shivenduverma-sap** [comment]
I didnt ask for changes? I already made the changes and now I have to adapt again?

**Comment by Skarlso** [comment]
I only added the local Tiltfile changes to help you out in adopting the local ocm-controller certificate generation. :)

File: e2e/main_test.go [hunk 7]
```diff
@@ -110,66 +99,6 @@ func checkIsNamespaceReady(name string) features.Func {
 	}
 }
 
-func replicateRegistryCerts(namespace string) features.Func {
```
Comment (shivenduverma-sap) [inline e2e/main_test.go]: This is required because ocirepositries in project namespace need the registry-secrets to access the registry
Reverting in the next push

Comment (Skarlso) [reply e2e/main_test.go]: It's not required because the Tiltfile is replicating the certificate.

Comment (Skarlso) [reply e2e/main_test.go]: Did you go over the changes I made? :)

Comment (shivenduverma-sap) [reply e2e/main_test.go]: You are relicating to the mpas-system namespace
they are also needed in the mpas-test-001 namespace, which the test will create

Comment (shivenduverma-sap) [reply e2e/main_test.go]: `apiVersion: source.toolkit.fluxcd.io/v1beta2
kind: OCIRepository
metadata:
  creationTimestamp: "2023-08-11T09:33:59Z"
  finalizers:
  - finalizers.fluxcd.io
  generation: 1
  name: backend-kustomization
  namespace: mpas-test-001
  ownerReferences:
  - apiVersion: delivery.ocm.software/v1alpha1
    kind: FluxDeployer
    name: backend-kustomization
    uid: 390e1b57-ac22-4368-a640-6f641aa812b4
  resourceVersion: "6106"
  uid: 3c5ffda1-a3dc-4616-bcb7-1d30c6231fc8
spec:
  certSecretRef:
    name: registry-certs
  interval: 0s
  provider: generic
  ref:
    tag: "3425"
  timeout: 60s
  url: oci://registry.ocm-system.svc.cluster.local:5000/sha-17168972171569427552
status:
  conditions:
  - lastTransitionTime: "2023-08-11T09:37:11Z"
    message: building artifact
    observedGeneration: 1
    reason: ProgressingWithRetry
    status: "True"
    type: Reconciling
  - lastTransitionTime: "2023-08-11T09:37:11Z"
    message: 'failed to generate transport for ''oci://registry.ocm-system.svc.cluster.local:5000/sha-17168972171569427552'':
      secrets "registry-certs" not found'
    observedGeneration: 1
    reason: AuthenticationFailed
    status: "False"
    type: Ready
  - lastTransitionTime: "2023-08-11T09:34:00Z"
    message: 'failed to generate transport for ''oci://registry.ocm-system.svc.cluster.local:5000/sha-17168972171569427552'':
      secrets "registry-certs" not found'
    observedGeneration: 1
    reason: AuthenticationFailed
    status: "True"
    type: FetchFailed
  observedGeneration: -1
`

Comment (Skarlso) [reply e2e/main_test.go]: Ah it's the project namespace. True 👍 Put it back in that case.

File: e2e/product_controller_test.go [hunk 8]
```diff
@@ -9,140 +8,196 @@ package e2e
 
 import (
 	"context"
+	"fmt"
+	corev1 "k8s.io/api/core/v1"
```
Comment (Skarlso) [inline e2e/product_controller_test.go:12]: Please sort these import statements according to:

- stdlib
- third party
- project based

Thanks.

File: e2e/product_controller_test.go [hunk 9]
```diff
@@ -151,105 +206,282 @@ func newProductFeature(mpasRepoName, mpasNamespace, projectRepoName string) *fea
...
+				t.Fatal(err)
 			}
-			fluxOci := &v1beta2.OCIRepository{
-				ObjectMeta: metav1.ObjectMeta{Name: "podinfo", Namespace: mpasNamespace},
+		}
+		return ctx
+	}
+}
+
+func checkKustomizationReady(namespace string, pipelineNames []string) features.Func {
```
Comment (Skarlso) [inline e2e/product_controller_test.go:372]: Looking really good. Would you please change these to also use the common ready function `checkObjectCondition`? Thanks.

Comment (shivenduverma-sap) [reply e2e/product_controller_test.go:372]: We are checking for 3, each separate call will return a status, will not be as pretty as it looks at first glance

Comment (Skarlso) [reply e2e/product_controller_test.go:372]: Understood. Change whatever you see fit / looks better. If not, feel free to leave it. 👍

**Comment by Skarlso** [comment]
@shivenduverma-sap adding release for the e2e framework https://github.com/open-component-model/ocm-e2e-framework/pull/35

**Comment by Skarlso** [comment]
ocm-e2e-framework v0.6.0 successfully released.

**Comment by Skarlso** [comment]
Needs:
```yaml
      - uses: yokawasa/action-setup-kube-tools@v0.9.2
        with:
          kustomize: '5.0.1'
          tilt: '0.32.2'
```

**Comment by shivenduverma-sap** [comment]
https://github.com/open-component-model/MPAS/blob/feature/56/e2e/suite_test.go#L46
Somehow Tilt is running before installing flux

**Comment by shivenduverma-sap** [comment]
@Skarlso Any clue on how to fix this?

**Comment by Skarlso** [comment]
Hmm, interesting. I will look at it later on. I'm off for now.

**Comment by Skarlso** [comment]
I'll look into this. It's super weird and doesn't make a lot of sense. :D

**Comment by Skarlso** [comment]
Okay, so for a first, the project controller was installing flux: https://github.com/open-component-model/mpas-project-controller/pull/31

**Comment by Skarlso** [comment]
```
Handling connection for 5000
--- PASS: TestMpasE2e/Setup_OCM_component_for_testing (0.72s)
=== RUN   TestMpasE2e/Configure_Management_Repository
=== RUN   TestMpasE2e/Configure_Management_Repository/management_namespace_mpas-system_exists
    env.go:422: checking if namespace with name: mpas-system exists
--- PASS: TestMpasE2e/Configure_Management_Repository/management_namespace_mpas-system_exists (5.00s)
--- PASS: TestMpasE2e/Configure_Management_Repository (10.02s)
=== RUN   TestMpasE2e/Create_Project
Handling connection for 3000

```

Things are working locally with `go work`. Not sure what the problem is on CI 🤔

**Comment by Skarlso** [comment]
AARGH. I don't fucking get this. I don't have these conflicts locally.

**Comment by Skarlso** [comment]
```
➜  MPAS-work   go work use ../MPAS
➜  MPAS-work go work use ../ocm-controller
➜  MPAS-work go work use ../git-controller
➜  MPAS-work go work use ../replication-controller
➜  MPAS-work go work use ../mpas-project-controller
➜  MPAS-work go work use ../mpas-product-controller
➜  MPAS-work go work sync
➜  MPAS-work

```

😭

**Comment by Skarlso** [comment]
I give up. No idea what the fuck is going on with this thing.

**Comment by Skarlso** [comment]
ah, I know where this is coming from

**Comment by Skarlso** [comment]
Okay, so what I didn't spot is OCM upgrading its dependency of k8s to the latest version. And since it's not using client-sdk it didn't care that client-sdk has breaking changes all over the place from 26-27.

**Comment by Skarlso** [comment]
We can choose to ignore it, or we adopt the new changes in all controllers.

**Comment by Skarlso** [comment]
I'm going to adopt it rather than choosing to lag behind. It's gonna take a little while to fix up all the controllers but they are small changes.

**Comment by Skarlso** [comment]
I have another idea... We are going to stop using `go work` in the end-to-end suit. Technically because of tilt it's no longer required. So it should be okay without it actually.

**Comment by Skarlso** [comment]
FINALLY!

```
Handling connection for 5000
--- PASS: TestMpasE2e/Setup_OCM_component_for_testing (0.45s)
=== RUN   TestMpasE2e/Configure_Management_Repository
=== RUN   TestMpasE2e/Configure_Management_Repository/management_namespace_mpas-system_exists
    env.go:422: checking if namespace with name: mpas-system exists
--- PASS: TestMpasE2e/Configure_Management_Repository/management_namespace_mpas-system_exists (5.00s)
--- PASS: TestMpasE2e/Configure_Management_Repository (10.01s)
=== RUN   TestMpasE2e/Create_Project
```

**Comment by Skarlso** [comment]
Looks like either the codeowners formatting is not working or gitea's file fetch strips formatting. 🤔 Going to investigate.

**Comment by Skarlso** [comment]
Screenshot 2023-08-14 at 9 05 13


Yep, formatting is off. I'm gonna fix that in git-controller.

**Comment by Skarlso** [comment]
ah it was only for gitea 🤦  I left out the break point.

**Comment by Skarlso** [comment]
Fixed the formatting on Gitea https://github.com/open-component-model/git-controller/pull/55

**Comment by shivenduverma-sap** [comment]
https://github.com/open-component-model/MPAS/blob/feature/56/.github/workflows/e2e.yaml#L99
Needs to be changed to `make e2e` before merge

**Comment by Skarlso** [comment]
And the branch to `main` on the MPAS checkout. :)

**Comment by shivenduverma-sap** [comment]
You can remove the ref for MPAS, it uses the current branch by default.

**Comment by Skarlso** [comment]
Ahhhh f*ck. That's going to be a problem
```
*** Test killed with quit: ran too long (11m0s).
```

**Comment by Skarlso** [comment]
Question is what killed it. Github, or go test. :)

**Comment by Skarlso** [comment]
ah it was go test. nice. I can just increase that.

**Review by Skarlso — APPROVED** [review]
